### PR TITLE
add Model - Model

### PR DESCRIPTION
**Pull request by Evilart86**
add CRUD base service logic
add CRUD base api logic
add CRUD generics "type not found" error
add base application model BaseModel

### CODE REVIEW

File: src/server/models/type.ts [hunk 1]
```diff
@@ -1,24 +1,7 @@
-import { Model, model, Schema, Document } from "mongoose";
+import { Model, model, Document } from "mongoose";
 import { Type } from "../interfaces/type";
+import TypeSchema from "../shemas/type";
```
Comment (aleh-rudzko) [inline src/server/models/type.ts]: @Evilart86 pls, rename "shemas" -> schemes

File: src/server/services/crud.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import { Model, Document } from "mongoose";
+import { CrudTypeNotFound } from "../errors";
+import BaseModel from "../interfaces/base";
+
+
+export interface ICrudService<TDocument extends Document & TIInstance, TIInstance extends BaseModel> {
```
Comment (aleh-rudzko) [inline src/server/services/crud.ts]: @Evilart86 What is TIInstance? can we simplify this name ? Also prefix I

File: src/tests/server/services/model.spec.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
+import { expect, should } from "chai";
+import { suite, test } from "mocha-typescript";
+import { ModelService } from "../../../server/services/model";
+import { ModelModel } from "../../../server/models/model";
+import IModel from "../../../server/interfaces/model";
+import BaseSpec from "./base.spec";
+
+@suite("Model service")
+class ModelSpec implements BaseSpec{
```
Comment (aleh-rudzko) [inline src/tests/server/services/model.spec.ts]: @Evilart86 pls, run gulp tslint and fix issues

File: src/server/errors.ts [hunk 4]
```diff
@@ -10,6 +10,13 @@ export class InternalServerError extends BaseError {
     }
 }
 
+export class CrudTypeNotFound<T> extends BaseError {
+    constructor() {
+        let ctor: { new (): T };
```
Comment (aleh-rudzko) [inline src/server/errors.ts]: @Evilart86 ctor -> constructor ? pls, don't use abbreviations in names. Also use **const** if variable not change

File: src/server/interfaces/model.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import BaseModel from "./base";
+
+export default interface IModel extends BaseModel {
```
Comment (aleh-rudzko) [inline src/server/interfaces/model.ts]: @Evilart86 https://github.com/basarat/typescript-book/blob/master/docs/styleguide/styleguide.md#interface
pls, don't use prefix I

File: src/server/interfaces/model.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+    id?: any;
+    name: string;
+    createdAt?: Date;
```
Comment (aleh-rudzko) [inline src/server/interfaces/model.ts]: @Evilart86 I think createdAt & updatedAt can be moved to BaseModel

File: src/server/models/model.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import { Model, model, Document } from "mongoose";
+import IModel from "../interfaces/model";
+import ModelSchema from "../shemas/model";
+
+
+export interface IModelDocument extends Document, IModel { }
```
Comment (aleh-rudzko) [inline src/server/models/model.ts]: @Evilart86 same prefix I

File: src/server/services/crud.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    create(instance: TIInstance): Promise<TIInstance>;
+
+    get(id: string): Promise<TIInstance>;
+
+    update(instance: TIInstance): Promise<TIInstance>;
+
+    remove(id: string): Promise<TIInstance>;
+}
+
+export interface ICrudMongooseService<TIInstance> extends ICrudService<Document & TIInstance, TIInstance> {}
```
Comment (aleh-rudzko) [inline src/server/services/crud.ts]: @Evilart86 prefix I

File: src/server/services/model.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import CrudService, { ICrudMongooseService } from "./crud";
+import { IModelDocument, ModelModel } from "../models/model";
+import IModel from "../interfaces/model";
+
+interface IModelService extends ICrudMongooseService<IModel> {}
```
Comment (aleh-rudzko) [inline src/server/services/model.ts]: @Evilart86 prefix I

**Review by aleh-rudzko — CHANGES_REQUESTED** [review]

**Comment by aleh-rudzko** [comment]
@Evilart86 Also don't forget about swagger

File: src/server/services/crud.ts (hunk 2 above)
Comment (Evilart86) [reply src/server/services/crud.ts]: Type of instance interface

File: src/server/errors.ts (hunk 4 above)
Comment (Evilart86) [reply src/server/errors.ts]: Ok, i'll rename it

File: src/server/interfaces/model.ts (hunk 5 above)
Comment (Evilart86) [reply src/server/interfaces/model.ts]: Why?

File: src/server/interfaces/model.ts (hunk 6 above)
Comment (Evilart86) [reply src/server/interfaces/model.ts]: Mb, Need to discuss

File: src/server/interfaces/model.ts (hunk 5 above)
Comment (aleh-rudzko) [reply src/server/interfaces/model.ts]: @Evilart86 Reason: Unconventional. lib.d.ts defines important interfaces without an I (e.g. Window, Document etc).

File: src/server/services/crud.ts (hunk 2 above)
Comment (aleh-rudzko) [reply src/server/services/crud.ts]: @Evilart86 remove prefix I

**Review by aleh-rudzko — APPROVED** [review]

**Pull request merged**